src/transformStateWithClones.js

```diff
@@ -7,7 +7,29 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
-}
+  let currentState = { ...state };
+  const stateHistory = [];
+
+  for (const action of actions) {
+    if (action.type === 'addProperties') {
+      currentState = { ...currentState, ...action.extraData };
+    }
+
+    if (action.type === 'removeProperties') {
+      currentState = { ...currentState };
+
+      for (const key of action.keysToRemove) {
+        delete currentState[key];
+      }
+    }
 
+    if (action.type === 'clear') {
+      currentState = {};
+    }
```

Comment:
Instead of using multiple `if` statements, consider using a `switch` statement to handle different types of actions. This will make your code cleaner and easier to understand, as well as ensure that you have a default case for error handling. Remember to include a default case in your switch statement as per the checklist item 3.